featurebyte/utils/async_helper.py

```diff
@@ -5,18 +5,137 @@
 from __future__ import annotations
 
 import asyncio
+import time
+import uuid
 from asyncio import Future
 from typing import Any, Coroutine, List
 
+from redis import Redis
 
-def asyncio_gather(*coros: Coroutine[Any, Any, Any], max_concurrency: int = 0) -> Future[List[Any]]:
+
+class RedisFairCountingSemaphore:
+    """
+    Fair counting semaphore using Redis adapted from "Redis In Action" by Josiah L. Carlson
+    """
+
+    def __init__(self, redis: Redis[Any], key: str, limit: int, timeout: int = 10):
+        """
+        Initialize semaphore
+
+        Parameters
+        ----------
+        redis: Redis[Any]
+            Redis connection
+        key: str
+            Semaphore key
+        limit: int
+            Semaphore limit count
+        timeout: int
+            Timeout in seconds
+        """
+        self._redis = redis
+        self._key = key
+        self._limit = limit
+        self._timeout = timeout
+        self._czset = f"{self._key}:owner"
+        self._ctr = f"{self._key}:counter"
+        self._lock = f"{self._key}:lock"
+        self._identifier: str | None = None
+
+    async def __aenter__(self) -> None:
+        """
+        Acquire semaphore on enter, blocks until semaphore is acquired or timeout
+
+        Raises
+        ------
+        TimeoutError
+            If semaphore is not acquired within timeout
+        """
+        # Acquire lock for semaphore
+        lock_timeout = 1
+        lock = self._redis.lock(self._lock, timeout=lock_timeout)
```

Comment:
From the documentation:
```
        If specified, ``timeout`` indicates a maximum life for the lock.
        By default, it will remain locked until release() is called.
```
Seems like the operation here is about increment the counter, 1s should be sufficient in this case.

Reply:
Looks like the lock should be released only after the semaphore is acquired, so it will need to be set to the same timeout as the semaphore

Reply:
Ok. If i understood it correctly, we only need 1 timeout parameter in this case.

Reply:
it seems so. @YS-L Would it make sense to set it to `LONG_RUNNING_EXECUTE_QUERY_TIMEOUT_SECONDS`?

Reply:
I think that makes sense. This lock expiring doesn't mean that the query will get to run immediately, but just to get a "ticket" to wait for the semaphore, is that right?

Reply:
From what I understand this timeout should start when the lock is successfully acquired. When it times out the lock is released for other clients even if it's not released by the lock owner. 

If we set this as long as the lock owner execution lifespan (or longer) this should keep the lock mechanism intact. If this is too short then another process may acquire the lock before it's released. If it's set too long the it will block other processes from acquiring the lock for an extended period if the lock owner is terminated in a way that does not release the lock